core/src/main/scala/chisel3/Aggregate.scala

```diff
@@ -1287,25 +1287,46 @@ abstract class Record extends Aggregate {
   // without having to recurse over all elements after the Record is
   // constructed. Laziness of _elements means that this check will
   // occur (only) at the first instance _elements is referenced.
-  // Also used to sanitize names and convert to more optimized VectorMap datastructure
+  // Also used to sanitize names and convert to more optimized VectorMap datastructure (if necessary)
+  // TODO We should figure out how to combine this with elements
+  // The main trick there is storing the sanitized names and making sure every Chisel API uses them
+  // Strawman proposal is to put sanitized name in the Slot refs of the children, and then replace
+  // _elements and elements with a single Array and deprecate elements as a public API
+  // This would also give an opportunity to stop storing elements in reverse order
   private[chisel3] lazy val _elements: VectorMap[String, Data] = {
     // Since elements is a map, it is impossible for two elements to have the same
     // identifier; however, Namespace sanitizes identifiers to make them legal for Firrtl/Verilog
     // which can cause collisions
     // Note that OpaqueTypes cannot have sanitization (the name of the element needs to stay empty)
     //   Use an empty Namespace to indicate OpaqueType
     val namespace = Option.when(!this._isOpaqueType)(Namespace.empty)
-    elements.view.map {
-      case (name, field) =>
-        if (field.binding.isDefined) {
-          throw RebindingException(
-            s"Cannot create Record ${this.className}; element ${field} of Record must be a Chisel type, not hardware."
-          )
-        }
-        // namespace.name also sanitizes for firrtl, leave name alone for OpaqueTypes
-        val sanitizedName = namespace.map(_.name(name, leadingDigitOk = true)).getOrElse(name)
-        sanitizedName -> field
-    }.to(VectorMap) // VectorMap has O(1) lookup whereas ListMap is O(n)
+    val originalElements = elements
+    // Don't create a new map unless necessary, this is much more memory efficient in common case
+    // This is true if elements is not a VectorMap or if any names need sanitization
+    var needNewMap = !originalElements.isInstanceOf[VectorMap[_, _]]
```

Comment:
is there a reason why this has to be done with a `var` and also why we aren't just doing it something like:

```
def makeNewMap(newNames: Seq[String) =       newNames.view
        .zip(originalElements)
        .map { case (name, (_, data)) => name -> data }
        .to(VectorMap) // VectorMap has O(1) lookup whereas ListMap is O(n)

val newElements = originalElements match {
  case original: VectorMap[_, _] => {
    val newNames = ...
    val allNamesUnchanged = newNames == origNames
    if (allNamesUnchanged) original else makeNewMap(newNames)
  case _ => makeNewMap(newNames)
}

...? I guess you are trying to only iterate over it once ...?

Reply:
> I guess you are trying to only iterate over it once ...?

Correct, I'm trying to minimize the total number of traversals